File: 2023/fredriv/day7/day7.scala
import scala.io.Source

case class Card(c: Char):
    def strength = c match {
        case 'A' => 14
        case 'K' => 13
        case 'Q' => 12
        case 'J' => 11
        case 'T' => 10
        case c => c - '0'
    }


case class Hand(cards: Seq[Card], bid: Int) extends Ordered[Hand]:
    override def toString = s"Hand(${cards.map(_.c).mkString}, $bid)"

    def strength =
        val numOfAKind = cards.groupBy(_.c).view.mapValues(_.size).values.toSeq
        numOfAKind.max match {
            case 5 => 7 // Five of a kind
            case 4 => 6 // Four of a kind
            case 3 =>
                if numOfAKind.contains(2) then 5 // Full house
                else 4 // Three of a kind
            case 2 =>
                if numOfAKind.filter(_ == 2).size == 2 then 3 // Two pairs
                else 2 // One pair
            case 1 => 1 // High card
        }

    def compare(that: Hand) =
        if this.strength == that.strength then
            this.cards.zip(that.cards).map(_.strength - _.strength).filterNot(_ == 0).head
        else this.strength - that.strength


@main def cards =
    val input = Source.fromFile("input.txt").getLines.toVector

    val hands =
        for s <- input
            Array(cards, bid) = s.split(" ")
        yield Hand(cards.map(Card), bid.toInt)

    val ranked = hands.sorted.zipWithIndex
    val scores = ranked.map((hand, idx) => hand.bid * (idx + 1))
    println("Part 1: " + scores.sum)

File: 2023/jakobgm/day7.py
from pathlib import Path
from typing import Counter


PROBLEM = """
32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483
""".strip()
PROBLEM = Path("input/7.txt").read_text().strip()

ORDERING = "AKQJT98765432"
STRENGTH = {card: strength for strength, card in enumerate(reversed(ORDERING))}


class Hand:
    def __init__(self, line: str) -> None:
        self.hand, bid = line.split()
        self.bid = int(bid)
        self.strengths = tuple(STRENGTH[card] for card in self.hand)

        count = Counter(self.hand)
        match list(sorted(count.values(), reverse=True)):
            case [5]:  # Full house
                self.kind = 6
            case [4, 1]:  # Four of a kind
                self.kind = 5
            case [3, 2]:  # Full house
                self.kind = 4
            case [3, 1, 1]:  # Three of a kind
                self.kind = 3
            case [2, 2, 1]:  # Two pairs
                self.kind = 2
            case [2, 1, 1, 1]:  # One pair
                self.kind = 1
            case [1, 1, 1, 1, 1]:  # High card
                self.kind = 0
            case _:
                raise RuntimeError()
        self.sort_key = (self.kind,) + self.strengths

    def __repr__(self) -> str:
        return f"{self.hand}"


hands = []
for line in PROBLEM.splitlines():
    hands.append(Hand(line=line))

hands.sort(key=lambda hand: hand.sort_key)
winnings = 0
for rank, hand in enumerate(hands):
    print(rank + 1, hand.bid, hand)
    winnings += (rank + 1) * hand.bid

print(winnings)


# --- Part two ---
NEW_ORDERING = "AKQT98765432J"
NEW_STRENGTH = {card: strength for strength, card in enumerate(reversed(NEW_ORDERING))}


class NewHand:
    def __init__(self, line: str) -> None:
        self.hand, bid = line.split()
        self.bid = int(bid)
        self.strengths = tuple(NEW_STRENGTH[card] for card in self.hand)

        count = Counter(self.hand.replace("J", ""))
        match list(sorted(count.values(), reverse=True)):
            case [] | [_]:  # Full house
                self.kind = 6
            case [_, 1]:  # Four of a kind
                self.kind = 5
            case [_, 2]:  # Full house
                self.kind = 4
            case [_, 1, 1]:  # Three of a kind
                self.kind = 3
            case [_, _, 1]:  # Two pairs
                self.kind = 2
            case [_, 1, 1, 1]:  # One pair
                self.kind = 1
            case [1, 1, 1, 1, 1]:  # High card
                self.kind = 0
            case _:
                raise RuntimeError()
        self.sort_key = (self.kind,) + self.strengths


new_hands = []
for line in PROBLEM.splitlines():
    new_hands.append(NewHand(line=line))

new_hands.sort(key=lambda hand: hand.sort_key)
new_winnings = 0
for rank, new_hand in enumerate(new_hands):
    new_winnings += (rank + 1) * new_hand.bid

print(new_winnings)
